File: viz/engine/shelby_viz/renderer.py
from __future__ import annotations

from pathlib import Path
from typing import Any

import jinja2

from .data import PipelinesContext, MockPipelinesContext, SeriesData
from .loader import DashboardDef, WidgetDef

BUILTINS_DIR = Path(__file__).parent / "builtins"

BUILTIN_TYPES = {"gauge", "timeseries", "counter", "table", "status"}


def _resolve_binding(cfg: Any, pipelines: PipelinesContext) -> Any:
    """Convert one binding config value into live data.

    Binding config shapes supported in dashboard.yml:

        # Structured (recommended) — pull a time series
        pipeline: my-slug
        series: field.path
        limit: 50          # optional
        unit: ms           # optional

        # Structured — pull run list
        pipeline: my-slug
        runs: 20           # limit

        # Static scalar
        value: 42

        # Jinja2 expression string  "{{ pipelines['slug'].series('field') }}"
        # Evaluated with eval() — power-user shorthand, runs locally so acceptable.
    """
    if isinstance(cfg, dict):
        if "series" in cfg:
            return pipelines[cfg["pipeline"]].series(
                cfg["series"],
                limit=cfg.get("limit", 50),
                unit=cfg.get("unit", ""),
            )
        if "runs" in cfg:
            return pipelines[cfg["pipeline"]].runs(limit=cfg.get("runs", 20))
        if "value" in cfg:
            return cfg["value"]
        return cfg

    if isinstance(cfg, str) and cfg.strip().startswith("{{"):
        # Strip {{ }} and eval the inner expression
        inner = cfg.strip()[2:-2].strip()
        try:
            return eval(inner, {"__builtins__": {}}, {"pipelines": pipelines})  # noqa: S307
        except Exception as exc:
            return SeriesData(kind="timeseries", unit="", points=[], meta={"error": str(exc)})

    return cfg


def _filesizeformat(value: float, binary: bool = False) -> str:
    base = 1024 if binary else 1000
    value = float(value)
    for unit in ["B", "KB", "MB", "GB", "TB"]:
        if abs(value) < base:
            return f"{value:.1f} {unit}"
        value /= base
    return f"{value:.1f} PB"


class Renderer:
    """Resolves bindings from the Shelby API and renders Jinja2 widget templates."""

    def __init__(
        self,
        dashboard_path: Path,
        shelby_url: str,
        pipelines: PipelinesContext | MockPipelinesContext | None = None,
    ) -> None:
        self.dashboard_path = Path(dashboard_path).resolve()
        self.shelby_url = shelby_url
        self._pipelines = pipelines
        self._env = self._build_env()

    def _build_env(self) -> jinja2.Environment:
        loaders = [
            # User's dashboard project directory (custom widgets, vars, assets)
            jinja2.FileSystemLoader(str(self.dashboard_path)),
            # Built-in widgets and shell shipped with shelby-viz
            jinja2.FileSystemLoader(str(BUILTINS_DIR)),
        ]
        env = jinja2.Environment(
            loader=jinja2.ChoiceLoader(loaders),
            autoescape=jinja2.select_autoescape(["html"]),
        )
        env.filters["filesizeformat"] = _filesizeformat
        env.filters["ratio"] = lambda v, total: v / total if total else 0
        env.filters["abs"] = abs
        return env

    def render_widget(self, widget: WidgetDef) -> str:
        """Fetch live data and render one widget to an HTML string."""
        pipelines = self._pipelines if self._pipelines is not None else PipelinesContext(self.shelby_url)

        data: dict[str, Any] = {}
        for key, binding_cfg in widget.binding.items():
            data[key] = _resolve_binding(binding_cfg, pipelines)

        if widget.type == "custom":
            template_path = widget.template
        else:
            template_path = f"widgets/{widget.type}.html.j2"

        tmpl = self._env.get_template(template_path)
        return tmpl.render(
            data=data,
            config=widget.config,
            meta={"title": widget.title, "id": widget.id},
        )

    def render_dashboard(self, dashboard: DashboardDef) -> str:
        """Render the full dashboard page with all widgets inlined."""
        initial: dict[str, str] = {}
        for widget in dashboard.widgets:
            try:
                initial[widget.id] = self.render_widget(widget)
            except Exception as exc:
                initial[widget.id] = (
                    f'<style>:host{{display:flex;align-items:center;justify-content:center;height:100%}}'
                    f'.err{{color:#f44336;font-size:.75rem;padding:8px;text-align:center}}</style>'
                    f'<div class="err">⚠ {exc}</div>'
                )

        tmpl = self._env.get_template("shell/base.html")
        return tmpl.render(
            dashboard=dashboard,
            initial_renders=initial,
            shelby_url=self.shelby_url,
        )

File: internal/viz/renderer.go
package viz

import (
	"embed"
	"fmt"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"strings"

	"github.com/flosch/pongo2/v6"
)

//go:embed builtins
var builtinsFS embed.FS

// LLMsText returns the contents of the embedded llms.txt file.
func LLMsText() ([]byte, error) {
	return builtinsFS.ReadFile("builtins/llms.txt")
}

// WidgetRender pairs a widget definition with its rendered HTML.
// Fields use names that pongo2 can resolve via strings.Title lookup:
//   - wr.widget  → Widget *WidgetDef
//   - wr.html    → Html string
type WidgetRender struct {
	Widget *WidgetDef
	Html   string
}

// combinedLoader implements pongo2.ILoader, searching the user directory
// first and falling back to embedded builtins.
type combinedLoader struct {
	userDir string
	fsys    embed.FS
}

func (l *combinedLoader) Abs(base, name string) string {
	if base == "" || strings.HasPrefix(name, "/") {
		return name
	}
	dir := filepath.Dir(base)
	joined := filepath.Join(dir, name)
	// Normalise to forward slashes for embed.FS compatibility.
	return filepath.ToSlash(joined)
}

func (l *combinedLoader) Get(path string) (io.Reader, error) {
	// Try the user directory first.
	if l.userDir != "" {
		full := filepath.Join(l.userDir, filepath.FromSlash(path))
		if f, err := os.Open(full); err == nil {
			return f, nil
		}
	}
	// Fall back to the embedded builtins FS.
	embPath := "builtins/" + strings.TrimPrefix(path, "/")
	f, err := l.fsys.Open(embPath)
	if err != nil {
		return nil, fmt.Errorf("viz: template not found: %s", path)
	}
	return f, nil
}

func init() {
	// Register custom pongo2 filters required by the builtin templates.

	pongo2.RegisterFilter("float", func(in *pongo2.Value, _ *pongo2.Value) (*pongo2.Value, *pongo2.Error) {
		return pongo2.AsValue(in.Float()), nil
	})

	pongo2.RegisterFilter("abs", func(in *pongo2.Value, _ *pongo2.Value) (*pongo2.Value, *pongo2.Error) {
		f := in.Float()
		if f < 0 {
			f = -f
		}
		return pongo2.AsValue(f), nil
	})

	pongo2.RegisterFilter("filesizeformat", func(in *pongo2.Value, _ *pongo2.Value) (*pongo2.Value, *pongo2.Error) {
		v := in.Float()
		for _, unit := range []string{"B", "KB", "MB", "GB", "TB"} {
			if v < 1000 {
				return pongo2.AsValue(fmt.Sprintf("%.1f %s", v, unit)), nil
			}
			v /= 1000
		}
		return pongo2.AsValue(fmt.Sprintf("%.1f PB", v)), nil
	})

	pongo2.RegisterFilter("ratio", func(in *pongo2.Value, param *pongo2.Value) (*pongo2.Value, *pongo2.Error) {
		total := param.Float()
		if total == 0 {
			return pongo2.AsValue(0.0), nil
		}
		return pongo2.AsValue(in.Float() / total), nil
	})
}

// Renderer resolves widget bindings and renders pongo2 templates.
type Renderer struct {
	dashboardPath string
	shelbyURL     string
	pipelines     *PipelinesContext
	mock          *MockPipelinesContext
	tplSet        *pongo2.TemplateSet
}

// NewRenderer creates a live renderer backed by a Shelby API server.
func NewRenderer(dashboardPath, shelbyURL string) *Renderer {
	r := &Renderer{
		dashboardPath: resolveDashboardDir(dashboardPath),
		shelbyURL:     shelbyURL,
		pipelines:     &PipelinesContext{BaseURL: shelbyURL},
	}
	r.tplSet = buildTemplateSet(r.dashboardPath)
	return r
}

// NewSandboxRenderer creates a renderer using mock data from sandbox.yml.
func NewSandboxRenderer(dashboardPath string, mock *MockPipelinesContext) *Renderer {
	r := &Renderer{
		dashboardPath: resolveDashboardDir(dashboardPath),
		mock:          mock,
	}
	r.tplSet = buildTemplateSet(r.dashboardPath)
	return r
}

func resolveDashboardDir(path string) string {
	if info, err := os.Stat(path); err == nil && !info.IsDir() {
		return filepath.Dir(path)
	}
	return path
}

func buildTemplateSet(userDir string) *pongo2.TemplateSet {
	loader := &combinedLoader{userDir: userDir, fsys: builtinsFS}
	return pongo2.NewSet("shelby-viz", loader)
}

// resolveBinding converts one binding config entry into live data.
func (r *Renderer) resolveBinding(cfg any) any {
	m, ok := cfg.(map[string]any)
	if !ok {
		return cfg
	}

	if _, hasSeries := m["series"]; hasSeries {
		slug, _ := m["pipeline"].(string)
		field, _ := m["series"].(string)
		limit := 50
		if v, ok := m["limit"]; ok {
			limit = int(toFloat64(v))
		}
		unit, _ := m["unit"].(string)
		if r.mock != nil {
			return r.mock.Get(slug).Series(field, limit, unit)
		}
		return r.pipelines.Get(slug).Series(field, limit, unit)
	}

	if runsVal, hasRuns := m["runs"]; hasRuns {
		slug, _ := m["pipeline"].(string)
		limit := 20
		if v, ok := runsVal.(int); ok {
			limit = v
		} else if v, ok := runsVal.(float64); ok {
			limit = int(v)
		}
		if r.mock != nil {
			return r.mock.Get(slug).Runs(limit)
		}
		return r.pipelines.Get(slug).Runs(limit)
	}

	if val, hasVal := m["value"]; hasVal {
		return val
	}

	return cfg
}

// RenderWidget fetches live data and renders one widget to an HTML string.
func (r *Renderer) RenderWidget(widget *WidgetDef) (string, error) {
	data := make(map[string]any, len(widget.Binding))
	for key, bindCfg := range widget.Binding {
		data[key] = r.resolveBinding(bindCfg)
	}

	templatePath := widget.Template
	if widget.Type != "custom" {
		templatePath = "widgets/" + widget.Type + ".html.j2"
	}

	tmpl, err := r.tplSet.FromCache(templatePath)
	if err != nil {
		return "", fmt.Errorf("viz: load template %s: %w", templatePath, err)
	}

	out, err := tmpl.Execute(pongo2.Context{
		"data":   data,
		"config": widget.Config,
		"meta": map[string]string{
			"title": widget.Title,
			"id":    widget.ID,
		},
	})
	if err != nil {
		return "", fmt.Errorf("viz: render widget %s: %w", widget.ID, err)
	}
	return out, nil
}

// RenderDashboard renders the full dashboard page with all widgets inlined.
func (r *Renderer) RenderDashboard(dashboard *DashboardDef, liveReload bool) (string, error) {
	renders := make([]WidgetRender, 0, len(dashboard.Widgets))
	for _, w := range dashboard.Widgets {
		html, err := r.RenderWidget(w)
		if err != nil {
			html = widgetError(err)
		}
		renders = append(renders, WidgetRender{Widget: w, Html: html})
	}

	tmpl, err := r.tplSet.FromCache("shell/base.html")
	if err != nil {
		return "", fmt.Errorf("viz: load base template: %w", err)
	}

	out, err := tmpl.Execute(pongo2.Context{
		"dashboard":      dashboard,
		"widget_renders": renders,
		"shelby_url":     r.shelbyURL,
		"live_reload":    liveReload,
	})
	if err != nil {
		return "", fmt.Errorf("viz: render dashboard: %w", err)
	}
	return out, nil
}

// ValidateTemplates checks that all widget templates exist and parse without error.
// Returns a list of (widgetID, templatePath, error) tuples for each problem found.
func (r *Renderer) ValidateTemplates(dashboard *DashboardDef) []string {
	var errs []string
	for _, w := range dashboard.Widgets {
		tpath := w.Template
		if w.Type != "custom" {
			tpath = "widgets/" + w.Type + ".html.j2"
		}
		// Check template is reachable.
		accessible := false
		if r.dashboardPath != "" {
			if _, err := os.Stat(filepath.Join(r.dashboardPath, filepath.FromSlash(tpath))); err == nil {
				accessible = true
			}
		}
		if !accessible {
			embPath := "builtins/" + strings.TrimPrefix(tpath, "/")
			if _, err := fs.Stat(builtinsFS, embPath); err == nil {
				accessible = true
			}
		}
		if !accessible {
			errs = append(errs, fmt.Sprintf("  %s (%s): template not found: %s", w.ID, w.Type, tpath))
			continue
		}
		if _, err := r.tplSet.FromFile(tpath); err != nil {
			errs = append(errs, fmt.Sprintf("  %s (%s): parse error: %v", w.ID, w.Type, err))
		}
	}
	return errs
}

func widgetError(err error) string {
	return fmt.Sprintf(
		`<style>:host{display:flex;align-items:center;justify-content:center;height:100%%}.err{color:#f44336;font-size:.75rem;padding:8px;text-align:center}</style><div class="err">%v</div>`,
		err,
	)
}
